### PR TITLE
Sm whitelist

### PR DESCRIPTION
**Pull request by motdotla**
Adds whitelisting

### CODE REVIEW

**Comment by motdotla** [comment]
@elbuo8 @devinus ^

File: lib/main.js [hunk 1]
```diff
@@ -18,43 +18,54 @@ function HTMLUglify(config) {
 
   var salt = config.salt || 'use the force harry';
```
Comment (elbuo8) [inline lib/main.js]: Unrelated, but, maybe use a random string every time?

File: lib/main.js [hunk 2]
```diff
@@ -18,43 +18,54 @@ function HTMLUglify(config) {
...
   return pointer;
 };
 
+HTMLUglify.prototype.isWhitelisted = function(type, value) {
+  if (type === 'class') {
+    value = '.' + value;
+  }
+  if (type === 'id') {
+    value = '#' + value;
+  }
```
Comment (elbuo8) [inline lib/main.js:47]: Use a `switch` maybe?

File: lib/main.js (hunk 1 above)
Comment (motdotla) [inline lib/main.js]: I like that idea. Recommended library for that? ObjectId?

Comment (devinus) [inline lib/main.js]: https://nodejs.org/api/crypto.html#crypto_crypto_pseudorandombytes_size_callback

Comment (elbuo8) [inline lib/main.js]: I would just use `Math.rand()`

Comment (devinus) [inline lib/main.js]: Needs a string.

File: lib/main.js [hunk 3]
```diff
@@ -68,32 +79,32 @@ HTMLUglify.prototype.populateLookups = function(type, value, pointer, lookups) {
...
 
   $('*[for]').each(function() {
-    self.pointerizeIdAndFor('for', $(this), lookups);
+    _this.pointerizeIdAndFor('for', $(this), lookups);
   });
 
   $('*[class]').each(function() {
-    self.pointerizeClass($(this), lookups);
+    _this.pointerizeClass($(this), lookups);
   });
```
Comment (elbuo8) [inline lib/main.js:101]: Any chance this can be done using `async.parallel` + `async.each`?

File: lib/main.js [hunk 4]
```diff
@@ -125,14 +136,14 @@ HTMLUglify.prototype.processRules = function(rules, lookups) {
 };
 
 HTMLUglify.prototype.rewriteStyles = function($, lookups) {
-  var self = this;
+  var _this = this;
 
   lookups = lookups || {};
 
   $('style').each(function() {
```
Comment (elbuo8) [inline lib/main.js:147]: `async.each`?

File: lib/main.js (hunk 3 above)
Comment (devinus) [inline lib/main.js:101]: Sure, but nothing in the library is async under the hood so it wouldn't impact anything.

File: lib/main.js (hunk 1 above)
Comment (motdotla) [inline lib/main.js]: I could easily convert `Math.random()` to a string. 

Thoughts on which would be the fastest? Math.random() or the crypto lib?

Currently this htmlUglify gets instantiated on each email sent.

**Comment by elbuo8** [comment]
@motdotla most of the comments I left are, I guess, unrelated to the PR. Go a head and merge :+1: 

I'd like to revisit this for perf reasons.

File: lib/main.js (hunk 1 above)
Comment (elbuo8) [inline lib/main.js]: I see.

`Math.rand() + ""`

:laughing:

File: lib/main.js (hunk 3 above)
Comment (motdotla) [inline lib/main.js:101]: Possibly, this each method is from cheerio here: https://github.com/cheeriojs/cheerio

I'm hesitant to go into that at the moment. My main goal is to provide whitelisting so I can fix slif-extension for Mark's process.

File: lib/main.js (hunk 1 above)
Comment (devinus) [inline lib/main.js]: Let's not be grody. 

https://github.com/ivanakimov/hashids.js/blob/master/lib/hashids.js#L289

`crypto.pseudoRandomBytes(ALPHABET.length).toString('utf8')`. Done.

File: lib/main.js (hunk 3 above)
Comment (devinus) [inline lib/main.js:101]: Unless you are calling out to a library implemented in C you cannot have a pure JavaScript async function, so converting these to use `async` would just be converting something that isn't async anyway (i.e. no benefit).

File: lib/main.js (hunk 1 above)
Comment (motdotla) [inline lib/main.js]: @devinus and @elbuo8 I ran a crude speed test. Math.random() is faster.

Results:

```
➜  ~  node speedtest.js
Execution took 894 ms
Execution took 55 ms
```

```
var crypto = require('crypto');

// exec code
var maximumCount = 100000;
var ts = +new Date;
for (i = 0; i < maximumCount; i++) { 
  var random = crypto.pseudoRandomBytes(256);
}

console.log('Execution took ' + ((+new Date) - ts) + ' ms')

var ts = +new Date;
for (i = 0; i < maximumCount; i++) { 
  var random = Math.random() + '';
}

console.log('Execution took ' + ((+new Date) - ts) + ' ms')
```

Comment (devinus) [inline lib/main.js]: Well yeah, Math.random uses no system entropy at all. Why do we care about which is faster? This happens once...

Comment (motdotla) [inline lib/main.js]: It gets instantiated every time an email is sent. See comment above ^

Comment (devinus) [inline lib/main.js]: It does? https://github.com/RebelMail/html-preflight/blob/master/lib/plugins/uglifier.js#L6

**Pull request merged**